apps/aecontract/src/aect_dispatch.erl

```diff
@@ -179,3 +191,45 @@ chain_state(#{ contract    := ContractPubKey
     aec_vm_chain:new_offchain_state(Trees, OnChainTrees, TxEnv,
                                     ContractPubKey).
 
+check_type_info(CallData, TypeInfo) ->
+    case aeso_data:from_binary({tuple, [typerep]}, CallData) of
+        {ok, {{tuple, [string, GivenArgType]} = Type}} ->
+            case aeso_data:from_binary({tuple, [typerep, Type]}, CallData) of
+                {ok, {_, {Function, Args}}} ->
+                    case aeso_compiler:arg_typerep_from_type_info(Function, TypeInfo) of
+                        {ok, GivenArgType} -> ok;
+                        {ok, ArgType} ->
+                            %% Check if both type reps give the same result
+                            ?DEBUG_LOG("Checking representation for ~p\n"
+                                       "Given   : ~p\n"
+                                       "Accepted: ~p"
+                                      , [Function, GivenArgType, ArgType]),
+                            ExpectedType = {tuple, [string, ArgType]},
+                            case aeso_data:from_binary({tuple, [typerep, ExpectedType]}, CallData) of
+                                {ok, {_, {_, Args}}} -> ok;
```

Comment:
Is this what we want? In particular it allows one to specify incorrect types of the constructors of a datatype that you aren't using. This is harmless, I suppose, but it's not clear why we'd want to allow it.